[WIP] Symfony console application

GodLesZ opened the pull request:
TODO:
- [x] Finish implementation
- [x] Update/fix existing tests
- [ ] Add more tests

GodLesZ commented on src/Command/ValidateCommand.php:
```diff
@@ -0,0 +1,66 @@
+        $installedPackages = $this->getInstalledPackages($directory);
+
+        $documentedDependencies = $this->parser->getDocumentedDependencies($filepath);
+        if ($documentedDependencies === null) {
```
TODO: Add error output

cheeZery commented on src/Command/ValidateCommand.php:
```diff
@@ -0,0 +1,66 @@
+        ));
+
+        foreach ($validationResult as $line) {
+            $output->writeln($line->toString());
```
Why do you use  ``$output`` in this case instead of ``$this->io``?

cheeZery commented on src/Command/BaseCommand.php:
```diff
@@ -0,0 +1,110 @@
+        $this->managerNode = new NodePackageManager();
+        $this->parser = new MarkdownParser();
+        $this->writer = new MarkdownWriter();
+        $this->validator = new PackageValidator();
```
Maybe we should consider only instantiating those dependencies needed for that command, e.g. UpdateCommand doesn't need Validator, but then again this might be the only thing so it might just be easier this way!

cheeZery commented on src/Dependencies/DependencyDataAdditionalContent.php:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * @return DependencyDataAdditionalContent
+     */
+    public function removeLasEmptyLine(): DependencyDataAdditionalContent
```
removeLas**t**EmptyLine

And didn't the Writer used to remove all empty lines at the end? But I guess it still does that in ``MarkdownWriter->cleanupAdditionalContent``, right?

cheeZery commented on src/Dependencies/DependencyList.php:
```diff
@@ -0,0 +1,118 @@
+     * The return value will be casted to boolean if non-boolean was returned.
+     * @since 5.0.0
+     */
+    public function offsetExists($offset): bool
```
There seems to be some copy pasta in here. Are this functions needed because you implemented ``\ArrayAccess``?

cheeZery commented on src/Parser/Exception/MissingFileException.php:
```diff
@@ -0,0 +1,11 @@
+    public function __construct($filepath)
+    {
+        parent::__construct(sprintf('File not found: %s', $filepath));
+    }
```
Maybe do the same for ``FailedToParseDependencyInformationException``

cheeZery commented on src/Writer/MarkdownWriter.php:
```diff
@@ -26,10 +29,9 @@ public function createDocumentation(string $filepath, array $installedPackages,
-                }
+                $documentedDependency = $documentedDependencies->get($packageManagerName, $name);
 
+                // @TODO
```
Is there something missing here? ;-) Do what...?

cheeZery commented on src/Writer/MarkdownWriter.php:
```diff
@@ -47,18 +49,20 @@ public function createDocumentation(string $filepath, array $installedPackages,
+        }
 
-            $handle = @fopen($filepath, "w");
+        // @TODO: Maybe add documentation for packages who were documented but not installed (anymore)
```
I guess I missspelled this, should probably be "packages *which* were documented"

cheeZery commented on src/Writer/MarkdownWriter.php:
```diff
@@ -47,18 +49,20 @@ public function createDocumentation(string $filepath, array $installedPackages,
+
+        foreach ($documentation as $line) {
+            // @TODO: which line break?!
+            fwrite($handle, "$line\r\n");
```
I know, not a concern of this issue, but can't we just use PHP_EOL here? Eh, there's probably a better way to write files anyway, we'll fix this later!

GodLesZ commented:
fyi i fixed all points and also added some CI tools. Just look on https://github.com/GodLesZ/depdoc/pull/1 and you will see some example output from travis & codecov.io.

To enable them on this repo just go to https://travis-ci.org/ and https://codecov.io/, login using your GitHub account and enable the service for this repo.
They should detect the `.travis-ci.yml` in this branch and start crawling.

If don't mind, I will delete the files and keep them in my private fork 🤣

GodLesZ commented:
I'll add some more tests to ensure the writer and validator are working like expected.

GodLesZ commented:
I've changed my email and GitHub now somehow brokes my reference. I'll fix that until tomorrow ^^

cheeZery commented:
So that's why you closed this PR? 🤔 
It looks like this thing is ready to be merged. Or do you still want to improve the testing?
And thanks for all the new stuff, interfaces everywhere, continous integration, etc.! 🙌 I often saw Travis CI in action, but never set it up nor configured it, so I probably should get to know it better and read some documentations. But first I will do as you told me after we merged this bad boy.

GodLesZ commented:
@cheeZery GitHub closed it because of my changed author. I'll fix it asap :sweat_smile:

cheeZery commented:
No worries, take your time ;-)
